### ISSUE TITLE
[Bug]: Order issues with Leads

### ISSUE THREAD

**Issue by kleintom** [SpeciesFileGroup/taxonworks#3877]
### Steps to reproduce the bug

```bash
Continuing the discussion from https://github.com/SpeciesFileGroup/taxonworks/commit/22a5a6c78216996f03647f6a61cb3c2fe4171071, I'll add more below. This is with the latest adjustments.
```

### Environment

Development (docker)

### Version

5cff93d3929c16f2934db34d7e0

**Comment by kleintom** [comment]
It looks like there are still ordering issues with leads with the latest changes. STR for `insert_couplet`:

1) Create a new couplet (on left or right side doesn't matter) and save some distinguishing text for left and right. Note in the database that the left has `position=1` and the right has `position=2`.
2) Go up a couplet and insert a new couplet for the couplet from step 1. Now note in the database that `position` for both nodes created in step 1 is equal to 1, and `position` of both nodes created during the insert is equal to 2. Sometimes (but not always?) the newly inserted node that carries the old children appears on the wrong side of the ui (it should appear on whatever side the insert occurred on).
3) Since `position` no longer distinguishes which lead of those pairs comes first, updates can now cause the ui/display order of those pairs to change. So, e.g. edit the left lead that you created in step 1 and click Update - now the "left" lead is on the right and the "right" is on the left in the ui. (It looks like the ui order switches only when you edit/update the left lead, presumably because since `position` no longer determines order now last_updated does, or something like that?) The same switching happens if you edit the inserted pair in the same way.

In my testing so far, the `Delete both sides and reparent...` button seems to be working fine.

**Comment by kleintom** [comment]
I'm all for getting this to work without using `numeric_order: true` so we don't have issues with db:create, but just to clarify, when we were using `has_closure_tree order: 'position', numeric_order: true, dont_order_roots: true` and has_closure_tree was computing `position`s for us, positions were only ever computed on child pairs (root nodes didn't have a `position` set as I recall, at any rate it definitely didn't increase or change with the number of trees in the table) and the child pairs all had `position` values of 0 or 1, so to my mind it acted as though `position` was scoped to `parent_id`, as it is now with `acts_as_list`.

**Comment by kleintom** [comment]
I probably won't have time to look at this tomorrow, but if nobody gets to it earlier I should be able to try some things this weekend.

 I'll note that yesterday and in the past I've been unable to reproduce some of these issues in specs. And I'm not sure, but the specs may be passing now by chance: we do check that children[0] and children[1] are what we expect after `insert_couplet` (i.e. that left and right are the correct nodes), but it may be working just by chance that children[1] was last updated after children[0] or something like that, so more adjustments can be made there to try to draw that out.

**Comment by mjy** [comment]
@kleintom I'd like to poke at this.  Did the PR resolve this in your mind?

**Comment by kleintom** [comment]
Yeah, I'm not a great fan of my initial reload-everything-and-set-positions-explicitly solution, but I think it does resolve the issue, and the last comment in the pr cleared up at least some of the confusion that had been in my mind concerning why things hadn't been working the way I had expected. I'd be grateful for more clarity though :-)

**Comment by mjy** [comment]
I've been poking at this code for a while just now.  Hope to get some comments shortly, but it's confusing me too.

**Comment by mjy** [comment]
@kleintom I don't have full satisfaction yet.  Pushed a new branch `lead_debug`.  Feel free to poke at it, I'll get back to it too.  If you flop the destroy/kids change I made back to what you had then you'll see 2 errors failing.  As is there are a number related to destroy.

**Comment by mjy** [comment]
@kleintom Spent (too much) more time on this.  I've merged some new code into development that resolves this and does a little more, see a644a205cb4261f71b9, however I feel I largely replicate your work and was infected with the same lack of full satisfication and clarity.  You'll see I reload, and actually re-save, the latter because I gave up.  I'm curious as to whether Rails 7 won't resolve some of this.

A couple of observations:
* I saw differences b/w `update(parent: obj)` and `(parent_id: obj.id)`, I don't think this is the root problem (consistent use throughout does not solve the issue in my testing, in fact it's the opposite).
* When in mx I stacked a lot of `expect()` in one `specify`.  I've learned this is isn't the best pattern to follow, in fact ideally we'd have 1 expectation per specify.  This is a performance hit, but when specs have many complicated inter-relationships this can help tease out specific problems.  I'd encourage you (and us all) to ignore my mx patterns and make specs very lean/flat
* I wrote `Lead.draw` to help visualize, this actually caught one `reload` requirement.  It might help you out in the future.
* I'm wondering if there are conflicting callbacks in acts_as_list and closure_tree.

Once again, thanks again for the work, logically really solid, etc. etc. I'm treating this as code-review of sorts and I'm very happy with how you layed it out.  The spec work let me iterate much much faster had reviewed with out.
